### PR TITLE
Added more details to query context to track query specific information including timers. Updated code to use TimerContext from query request to record times  for different phases of query.

### PR DESCRIPTION
**Pull request by antumbde**

### CODE REVIEW

File: pinot-common/src/main/java/com/linkedin/pinot/common/query/QueryRequest.java [hunk 1]
```diff
@@ -30,24 +37,88 @@
...
+  public QueryRequest(InstanceRequest request, ServerMetrics serverMetrics) {
     this.instanceRequest = request;
+    this.serverMetrics = serverMetrics;
+    BrokerRequest brokerRequest = (request == null) ? null : request.getQuery();
+    timerContext = new TimerContext(brokerRequest, serverMetrics);
   }
 
-  public String getClientId() {
-    return clientId;
+  public @Nullable String getClientId() {
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/com/linkedin/pinot/common/query/QueryRequest.java]: why call this getClientId() ?  Why not getBrokerId() ?

File: pinot-common/src/main/java/com/linkedin/pinot/common/query/context/TimerContext.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    Timer(ServerQueryPhase phase) {
+      this.queryPhase = phase;
+    }
+
+    void start() {
+      startTime = System.nanoTime();
+      isRunning = true;
+    }
+
+    void setStartTime(long startTime) {
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/com/linkedin/pinot/common/query/context/TimerContext.java]: setStartTimeNs( long startTimeNs)
makes it more readable to have the unit in the method, thanks

File: pinot-common/src/main/java/com/linkedin/pinot/common/query/context/TimerContext.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    public void stopAndRecord() {
+      if (isRunning) {
+        endTime = System.nanoTime();
+
+        recordPhaseTime(this);
+        isRunning = false;
+      }
+    }
+
+    public long getDuration() {
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/com/linkedin/pinot/common/query/context/TimerContext.java]: please add units in the method name

File: pinot-common/src/main/java/com/linkedin/pinot/common/query/context/TimerContext.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+      stopAndRecord();
+    }
+  }
+
+  public TimerContext(BrokerRequest brokerRequest, ServerMetrics serverMetrics) {
+    this.brokerRequest = brokerRequest;
+    this.serverMetrics = serverMetrics;
+  }
+
+  public void setQueryArrivalTime(long queryStartTime) {
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/com/linkedin/pinot/common/query/context/TimerContext.java]: pls add units in method name and arg name

File: pinot-common/src/main/java/com/linkedin/pinot/common/query/context/TimerContext.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+  void recordPhaseTime(Timer phaseTimer ) {
+    serverMetrics.addPhaseTiming(brokerRequest, phaseTimer.queryPhase, phaseTimer.getDuration());
+    phaseTimers.remove(phaseTimer.queryPhase);
+  }
+
+  // for logging
+  @Override
+  public String toString() {
+    return String.format("%d,%d,%d,%d,%d,%d,%d,%d", queryArrivalTimeNanos,
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/com/linkedin/pinot/common/query/context/TimerContext.java]: so these get printed in nanos, right?

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/executor/ServerQueryExecutorV1Impl.java [hunk 6]
```diff
@@ -125,21 +127,24 @@ public DataTable processQuery(final QueryRequest queryRequest) {
...
       instanceResponse = globalQueryPlan.getInstanceResponse();
-      final long end = System.currentTimeMillis();
+      queryProcessingTimer.stopAndRecord();
+
       LOGGER.debug("Searching Instance for Request Id - {}, browse took: {}", instanceRequest.getRequestId(),
-          (end - start));
+          queryProcessingTimer.getDuration());
       LOGGER.debug("InstanceResponse for Request Id - {} : {}", instanceRequest.getRequestId(), instanceResponse.toString());
-      instanceResponse.getMetadata().put("timeUsedMs", Long.toString((end - start)));
+      instanceResponse.getMetadata().put("timeUsedMs", Long.toString(queryProcessingTimer.getDuration()));
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/executor/ServerQueryExecutorV1Impl.java]: are we changing units from ms to ns here? Not sure

File: pinot-core/src/main/java/com/linkedin/pinot/core/query/executor/ServerQueryExecutorV1Impl.java [hunk 7]
```diff
@@ -148,10 +153,11 @@ public DataTable processQuery(final QueryRequest queryRequest) {
       instanceResponse = new DataTable();
       instanceResponse.addException(QueryException.getException(QueryException.QUERY_EXECUTION_ERROR, e));
       TraceContext.logException("ServerQueryExecutorV1Impl", "Exception occurs in processQuery");
-      long end = System.currentTimeMillis();
-      LOGGER.info("Searching Instance for Request Id - {}, browse took: {}", requestId, requestId, (end - start));
+      queryProcessingTimer.stopAndRecord();
+
+      LOGGER.info("Searching Instance for Request Id - {}, browse took: {}", requestId, requestId, queryProcessingTimer.getDuration());
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/executor/ServerQueryExecutorV1Impl.java]: one info log per query please

Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/executor/ServerQueryExecutorV1Impl.java]: that's existing code not new statement. 

That said, the log statement and incorrect so I'll fix it

Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/executor/ServerQueryExecutorV1Impl.java]: Hmm.. the existing one is a debug (in the source tree). Not sure what is going on here.

Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/query/executor/ServerQueryExecutorV1Impl.java]: Ah wait. My bad. 

There are two statements, this one is in the exception path. If so, please leave it as info, thanks,

File: pinot-common/src/main/java/com/linkedin/pinot/common/query/QueryRequest.java (hunk 1 above)
Comment (antumbde) [inline pinot-common/src/main/java/com/linkedin/pinot/common/query/QueryRequest.java]: changed to getBrokerId()
Initial thinking was to keep it more generic (for non-broker clients sending requests) but that can be confusing

File: pinot-common/src/main/java/com/linkedin/pinot/common/query/context/TimerContext.java (hunk 5 above)
Comment (antumbde) [inline pinot-common/src/main/java/com/linkedin/pinot/common/query/context/TimerContext.java]: yes, added unit in the string format for clarity

**Pull request merged**